Step 3 of DGE analysis of matched polya and stranded samples

logstar opened the pull request:
This is the third one of the four pull requests for creating the PediatricOpenTargets/OpenPBTA-analysis/analysis/rna-seq-protocol-dge analysis module. These four pull requests correspond to the following four analysis steps:

1. Collapse pbta-gene-counts-rsem-expected_count matrices of stranded and poly-A libraries using the OpenPBTA-analysis/analyses/collapse-rnaseq analysis module.
2. Generate a HTML report of the collapsed genes using the OpenPBTA-analysis/analyses/collapse-rnaseq analysis module.
3. **Run edgeR exactTest, glmLRT, and glmQLFTest comparing stranded and poly-A RNA-seq rsem-expected_counts, using an adapted workflow from the OMPARE project.**
4. Select genes that are stably expressed in poly-A and stranded RNA-seq libraries.

The overall purposes of the PediatricOpenTargets/OpenPBTA-analysis/analysis/rna-seq-protocol-dge module are:

> - Check if the DGE analysis between polya and stranded data follow a null-p-value distribution.
> - Determine stably expressed transcripts between polyA and stranded samples.

### Purpose/implementation Section

#### What scientific question is your analysis addressing?

Check if the DGE analysis between polya and stranded data follow a null-p-value distribution.

#### What was your approach?

1. Select `Kids_First_Biospecimen`s with the same `sample_id`s and both stranded and poly-A `RNA_library`s.
    Index | Kids_First_Biospecimen_ID | sample_id | experimental_strategy | RNA_library | cohort
    ------|---------------------------|-----------|-----------------------|-------------|--------
    1     | BS_HE0WJRW6               | 7316-1455 | RNA-Seq               | stranded    | CBTN
    2     | BS_HWGWYCY7               | 7316-1455 | RNA-Seq               | poly-A      | CBTN
    3     | BS_SHJA4MR0               | 7316-161  | RNA-Seq               | stranded    | CBTN
    4     | BS_X0XXN9BK               | 7316-161  | RNA-Seq               | poly-A      | CBTN
    5     | BS_FN07P04C               | 7316-255  | RNA-Seq               | stranded    | CBTN
    6     | BS_W4H1D4Y6               | 7316-255  | RNA-Seq               | poly-A      | CBTN
    7     | BS_8QB4S4VA               | 7316-536  | RNA-Seq               | stranded    | CBTN
    8     | BS_QKT3TJVK               | 7316-536  | RNA-Seq               | poly-A      | CBTN
    9     | BS_7WM3MNZ0               | A16915    | RNA-Seq               | poly-A      | PNOC003
    10    | BS_KABQQA0T               | A16915    | RNA-Seq               | stranded    | PNOC003
    11    | BS_68KX6A42               | A18777    | RNA-Seq               | poly-A      | PNOC003
    12    | BS_D7XRFE0R               | A18777    | RNA-Seq               | stranded    | PNOC003
2. Run edgeR `exactTest`, `glmLRT`, and `glmQLFTest` comparing stranded and poly-A RNA-seq `rsem-expected_count`s, using an adapted workflow from the [OMPARE project](https://github.com/d3b-center/OMPARE/blob/master/code/patient_level_analyses/utils/rnaseq_edger_normalizations.R#L37-L41). RSEM expected read counts are normalized using the UQ-pgQ2 method, because UQ-pgQ2 normalization method has lower false positive rate in DGE testing (<https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-020-6502-7>), as suggested by @aadamk at PediatricOpenTargets ticket 17 <https://github.com/PediatricOpenTargets/ticket-tracker/issues/17#issuecomment-847007752>. The edgeR `common.dispersion` is estimated using the house keeping gene table `input/Housekeeping_GenesHuman.csv` downloaded from [Housekeeping Transcript Atlas](https://housekeeping.unicamp.br).

#### What GitHub issue does your pull request address?

<https://github.com/PediatricOpenTargets/ticket-tracker/issues/17>

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

The `03-edger-protocol-dge.R` is adapted from the [OMPARE project](https://github.com/d3b-center/OMPARE/blob/2c5cb9f56dc95212f113a6d13459e86b9d70c48f/code/patient_level_analyses/utils/rnaseq_edger_normalizations.R) to run edgeR `exactTest`, `glmLRT`, and `glmQLFTest` on UQ-pgQ2 normalized counts.

Are the `group` and `design` matrix defined correctly for the DGE tests? Relevant lines pasted below:

- `03-edger-protocol-dge.R` line 28-37:

```R
counts <- cbind(
    stranded[common_gene_id, c("BS_HE0WJRW6", "BS_SHJA4MR0", "BS_FN07P04C",
                               "BS_8QB4S4VA", "BS_KABQQA0T", "BS_D7XRFE0R")],
    polya[common_gene_id, c("BS_HWGWYCY7", "BS_X0XXN9BK", "BS_W4H1D4Y6",
                            "BS_QKT3TJVK", "BS_7WM3MNZ0", "BS_68KX6A42")]
)
group <- factor(c(rep("stranded", 6), rep("polya", 6)))

### build uq.pgQ2 normalized DGEList
counts_object = DGEList(counts = counts, group = group)
```

- `03-edger-protocol-dge.R` line 88:

```R
design <- model.matrix(~group)
```

Is the `common.dispersion` estimated correctly? Relevant `03-edger-protocol-dge.R` lines 75-86 pasted below:

```R
### Estimate common dispersion from a set of housekeeping genes.
hkgenes <- read.csv(file.path('input', 'Housekeeping_GenesHuman.csv'),
                    sep = ";", stringsAsFactors = FALSE)
hkgenes <- unique(hkgenes$Gene.name)
hkgenes <- intersect(rownames(counts_object), hkgenes)
hkg_counts_object <- counts_object[hkgenes, ]
# the expression prints "Design matrix not provided. Switch to the classic 
# mode.". Suppress this message when running.
out_msg <- capture.output(
    hkg_counts_object <- estimateDisp(hkg_counts_object, trend.method="none",
                                      tagwise=FALSE))
counts_object$common.dispersion <- hkg_counts_object$common.dispersion
```

Are the DGE tests called correctly? Relevant `03-edger-protocol-dge.R` lines pasted below:

```R
# Line 92-93
fit <- glmFit(counts_object, design)
lrt <- glmLRT(fit)
# Line 132
et <- exactTest(counts_object)
# Line 168-169:
ql_fit <- glmQLFit(counts_object, design)
qlft <- glmQLFTest(ql_fit)
```

#### Is there anything that you want to discuss further?

The current analysis script `03-edger-protocol-dge.R` hardcodes sample IDs, in order to simplify the procedure to find the biospecimens with the same `sample_ID` and both `poly-A` and `stranded` RNA-seq libraries.

However, it is discouraged to hardcode sample IDs in analysis scripts, partially because the continuous integration (CI) datasets have much fewer samples for testing.

If it is necessary remove the hardcoded sample IDs, I could write a subroutine to find the biospecimens to construct the `counts_object`, which might take one or two days of work. Even after this, the CI will not be able to run if the required biospecimens are not in the CI datasets.

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes.

### Results

#### What types of results are included (e.g., table, figure)?

- Tables:
    - `results/stranded_vs_polya_dge_exact_test_res.csv`: edgeR exactTest result table
    - `results/stranded_vs_polya_dge_lrt_res.csv`: edgeR LRT result table
    - `results/stranded_vs_polya_dge_ql_ftest_res.csv`: edgeR QL F-test result table
- Figures:
    - `plots/stranded_vs_polya_dge_exact_test_pvals_histogram.png`: histogram of edgeR exactTest p-values
    - `plots/stranded_vs_polya_dge_lrt_pvals_histogram.png`: histogram of edgeR LRT p-values
    - `plots/stranded_vs_polya_dge_ql_ftest_pvals_histogram.png`: histogram of edgeR QL F-test p-values

#### What is your summary of the results?

The results are summarized in [`README.md`](https://github.com/logstar/OpenPBTA-analysis/tree/rna-seq-protocol-dge-third/analyses/rna-seq-protocol-dge).

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [x] This analysis module has a `README` and it is up to date.
- [x] The analytical code is documented and contains comments.

aadamk commented:
Hi @logstar, thank you for these analyses. 

To answer your questions above: 

> Are the group and design matrix defined correctly for the DGE tests? 

Yes, they have been assigned correctly for this particular analysis. To partially answer your question regarding hardcoding, a more generalized routine for pulling and assigning subjects to groups will likely be preferred (though I'm not sure as to whether the plan is to make this a one-time analysis or a more generic module).  

> Is the common.dispersion estimated correctly? 

Tagwise or trended dispersions are usually the preferred method, as common dispersion does not capture mean-variance relationships. 
To answer this fully, we would need 

`estimateDisp(y, design=design, prior.df=NULL, trend.method="locfit", tagwise=TRUE)`
`estimateDisp(y, design=design, prior.df=NULL, trend.method=“none”)`
`plotBCV(y)`

Then, we would need to verify that common, trended, and tag wise dispersions have been estimated with the above function calls (in the DGElist object), and plotting the bcv would aid in determining the optimal fit. If there is a strong mean-dispersion trend, the common dispersion would be unsuitable. If the gene-level dispersions fluctuate around the trend, the trended dispersion is won't work and the tagwise dispersion would be required. Note: only trended dispersion is used under the QL pipeline.

> Are the DGE tests called correctly?

DGE tests are called correctly

A couple of additional notes:
1. @taylordm noted yesterday afternoon that DESeq2 is more widely used, so likely a "Part 5" PR for this module would involve leveraging their normalization + dge analysis with this same dataset as an additional benchmark (will also likely require a separate ticket).
2. p-value distributions would be expected to be non-uniform for this first-pass DGE, as the library enrichment method is the source of variation and rightward skewing seen in the current histograms. After the selection of empirically defined negative control genes and using those for batch correction, we'll more likely see a uniform distribution (which is true under the null hypothesis of no differences between groups). 
3. Regarding normalization, my intention had been to evaluate UQ-pgQ2 alongside edgeR's native TMM normalization (sorry if my comment in the ticket was unclear). Can you include a TMM normalized analysis in parallel to this (cc @jharenza to inform on folder structure)?

logstar commented:
@aadamk Thank you for the detailed review.

I will make the following updates:

- Revise dispersion estimation procedure.
- Run the same analysis with TMM normalization. Sorry that I misunderstood your comment on the normalization procedure. I removed TMM normalization, as I saw UQ-pgQ2 has an overall lower false positive rate in the Figure 1 of <https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-020-6502-7> . I agree it would be more informative to compare TMM and UQ-pgQ2 normalization methods on our own datasets.
- Run DESeq2 for DGE analysis.
- Follow-up with the discussion on whether this analysis should be a one-time analysis or generic.

logstar commented:
Following are the updates on this PR:

**Use trended tagwise dispersion estimation, as mean-dispersion trends were observed in edgeR BCV plots.**

TMM normalized `plotBCV`:

![tmm_bcv](https://raw.githubusercontent.com/logstar/OpenPedCan-analysis/efa81a1c3868755e67a356067748c28d0b8b4bf2/analyses/rna-seq-protocol-dge/plots/tmm_normalized/estimated_dispersions.png)

UQ-pgQ2 normalized `plotBCV`:

![up_bcv](https://raw.githubusercontent.com/logstar/OpenPedCan-analysis/efa81a1c3868755e67a356067748c28d0b8b4bf2/analyses/rna-seq-protocol-dge/plots/uqpgq2_normalized/estimated_dispersions.png)

`03-protocol-dge-seg.R` code for dispersion estimation:

```R
counts_object <- estimateDisp(
    counts_object, design=design, prior.df=NULL, trend.method='locfit',
    tagwise=TRUE)
```

**Add TMM normalization DGE analysis.**

`03-protocol-dge-seg.R` takes a `normalization` argument to determine whether use TMM or UQ-pgQ2 for normalization. `03-protocol-dge-seg.R` code for normalization:

```R
counts_object <- DGEList(counts = counts, group = group)
counts_object <- counts_object[
    filterByExpr(counts_object), , keep.lib.sizes=FALSE]

if (identical(norm_method, 'tmm')) {
    counts_object <- calcNormFactors(counts_object, method="TMM")
    norm_cnt_mat <- cpm(counts_object)
    round_norm_cnt_mat <- round(norm_cnt_mat)
    plot_nm_str <- 'TMM'
} else if (identical(norm_method, 'uqpgq2')) {
    counts_object <- DGEList(counts = uq.pgQ2(X = counts_object$counts),
                            group = group)
    norm_cnt_mat <- counts_object$counts
    round_norm_cnt_mat <- round(norm_cnt_mat)
    counts_object <- calcNormFactors(counts_object, method="none")
    plot_nm_str <- 'UQ-pgQ2'
}
```

**Add DESeq2 DGE analyses, either using TMM/UQ-pgQ2 normalized counts or DESeq2 standard DGE workflow.**

`03-protocol-dge-seg.R` runs DESeq2 DGE analysis using normalized read counts, with the following code:

```R
suppressMessages(
    dds <- DESeq2::DESeqDataSetFromMatrix(
        countData = round_norm_cnt_mat,
        colData = data.frame(group),
        design = ~ group)
)
DESeq2::sizeFactors(dds) <- rep(1, dim(round_norm_cnt_mat)[2])
suppressMessages(dds <- DESeq2::estimateDispersions(dds, fitType='local'))
dds <- DESeq2::nbinomWaldTest(dds)
# DESeq2::counts(dds, normalized=TRUE)[1:5, 1:5]
deseq2_res <- DESeq2::results(dds, cooksCutoff = FALSE, pAdjustMethod = 'BH')
```

The code is adapted from the shared code of [Li X, Cooper NGF, O’Toole TE, Rouchka EC. Choice of library size normalization and statistical methods for differential gene expression analysis in balanced two-group comparisons for RNA-seq studies. BMC Genomics. 2020;21: 75. doi:10.1186/s12864-020-6502-7](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-020-6502-7#MOESM3)

`04-deseq2-protocol-dge.R` runs DESeq2 standard DGE analysis on RSEM expected counts, with the following code:

```R
# DESeq2 requires integer matrix as read count matrix
round_cnt_mat <- round(counts_object$counts)
suppressMessages(
  dds <- DESeq2::DESeqDataSetFromMatrix(
    countData = round_cnt_mat,
    colData = data.frame(group),
    design = ~ group)
)
# From DESeq2 documentation:
# DESeq2 performs a default analysis through the steps:
# - estimation of size factors: estimateSizeFactors
# - estimation of dispersion: estimateDispersions
# - Negative Binomial GLM fitting and Wald statistics: nbinomWaldTest
suppressMessages(
  dds <- DESeq2::DESeq(dds)
)
```

aadamk commented on analyses/rna-seq-protocol-dge/03-protocol-dge-seg.R:
```diff
@@ -0,0 +1,315 @@
+ggsave(file.path(plot_outdir,
+                 'stranded_vs_polya_dge_ql_ftest_pvals_histogram.png'),
+       dpi = 300, plot = p, width = 8, height = 7)
+#------------ Run DESeq2 nbinomWaldTest ------------------------
```
Is this section from lines 255-315 distinct from `04-deseq2-protocol-dge.R`? Looks as though the only variation is the non-default estimation of dispersions.

logstar replied:
Sorry for the confusion. This section is distinct from `04-deseq2-protocol-dge.R`.

In addition to the differences on dispersion estimation methods, the normalization methods are also different. `03-protocol-dge-seg.R` uses TMM normalized CPM or UQ-pgQ2 normalized read counts for DESeq2 DGE analysis, whereas `04-deseq2-protocol-dge.R` uses the DESeq2 default normalization Relative Log Expression (RLE) method.

aadamk commented on analyses/rna-seq-protocol-dge/README.md:
```diff
@@ -0,0 +1,192 @@
+
+### Purpose
+
+Perform differential gene expression (DGE) analysis to compare RNA-seq libraries that are prepared using poly-A or ribodeplete-stranded protocols from the same samples.
```
```suggestion
Perform a first-pass differential gene expression (DGE) analysis (as described in the [RUVSeq vignette](https://bioconductor.riken.jp/packages/3.0/bioc/vignettes/RUVSeq/inst/doc/RUVSeq.pdf)) among samples prepared using both poly-A and ribodeplete-stranded protocols. Using housekeeping genes identified as part of the Housekeeping Transcript Atlas, the goal of first-pass DGE is to identify empirical negative control genes in a paired poly-A vs ribo-deplete comparison to identify stably expressed genes that would be suitable to inform batch corrections for other DGE comparisons whose samples vary by RNA-seq library enrichment method. 
```

aadamk commented on analyses/rna-seq-protocol-dge/README.md:
```diff
@@ -0,0 +1,192 @@
+
+Perform differential gene expression (DGE) analysis to compare RNA-seq libraries that are prepared using poly-A or ribodeplete-stranded protocols from the same samples.
+
+Check whether the p-values from DGE analysis follow a uniform distribution.
```
```suggestion
Evaluate the p-value distribution from first-pass DGE analysis, which will likely exhibit rightward skew.
```

aadamk commented on analyses/rna-seq-protocol-dge/README.md:
```diff
@@ -0,0 +1,192 @@
+
+Check whether the p-values from DGE analysis follow a uniform distribution.
+
+Select genes that are stably expressed in poly-A and stranded RNA-seq libraries.
```
```suggestion
Select genes that are stably expressed in poly-A and stranded RNA-seq libraries as empirical negative control genes for use in other DGE analyses varying by library enrichment method.
```

aadamk commented on analyses/rna-seq-protocol-dge/README.md:
```diff
@@ -0,0 +1,192 @@
+### Purpose
+
+Perform differential gene expression (DGE) analysis to compare RNA-seq libraries that are prepared using poly-A or ribodeplete-stranded protocols from the same samples.
+
```
```suggestion
Housekeeping gene list (HRT Atlas v1.0 database; PMID: 32663312) - https://housekeeping.unicamp.br/?download
```

aadamk commented on analyses/rna-seq-protocol-dge/03-protocol-dge-seg.R:
```diff
@@ -0,0 +1,315 @@
+    counts_object <- calcNormFactors(counts_object, method="none")
+    plot_nm_str <- 'UQ-pgQ2'
+}
+### Estimate dispersion
```
Seems the only step missing for all DGE analyses (edger and deseq2) in this module is subsetting down to housekeeping genes (per this ticket comment (https://github.com/PediatricOpenTargets/ticket-tracker/issues/17#issue-896875722)). 
So there would only be one modification to the current steps (step 3): 
1. Filter out non-expressed genes on the full transcriptome
2. Correct for compositional biases on the full transcriptome (TMM, uqpgq2, or DESeq2's method). 
3. _Filter down to housekeeping genes from https://housekeeping.unicamp.br/?download (Full Housekeeping gene list, ~2000 genes)_
4. Estimate dispersions
5. Run DGE and plot p-val histograms.

aadamk reviewed (changes requested):
Thank you for these detailed analyses @logstar 

There is only one key step missing for all DGE analyses performed here, which is subsetting down to housekeeping genes prior to running DGE. I've written a more detailed comment in `03-protocol-dge-seg.R` that shows at what point subsetting should be performed (after filtering by expression and correcting for lib size, but before estimating dispersions and running dge). 

Once that step is added, we can then select the negative control genes from the preferred method.

logstar commented:
@aadamk Thank you for the detailed review.

I understand the proposed modifications to run DGE analysis only on housekeeping genes, but I would like to confirm that this is requested in <https://github.com/PediatricOpenTargets/ticket-tracker/issues/17#issue-896875722>. In the following Krutika's code, the housekeeping genes are used to estimate common dispersion, whereas all genes are used for DGE analysis.

```R
### Copy normalized DGEList object (new object counts_object_1) and replace group variable with only one treatment group
counts_object_1=counts_object
counts_object_1$samples$group <- 1

### Estimate common dispersion from a set of housekeeping genes (linked below). 
y0 <- estimateDisp(counts_object_1[housekeeping,], trend="none", tagwise=FALSE)
counts_object$common.dispersion <- y0$common.dispersion

### Make design matrix 
design <- model.matrix(~Groups)
fit <- glmFit(counts_object, design)
lrt <- glmLRT(fit)
```

aadamk commented:
> @aadamk Thank you for the detailed review.
> 
> I understand the proposed modifications to run DGE analysis only on housekeeping genes, but I would like to confirm that this is requested in [PediatricOpenTargets/ticket-tracker#17 (comment)](https://github.com/PediatricOpenTargets/ticket-tracker/issues/17#issue-896875722). In the following Krutika's code, the housekeeping genes are used to estimate common dispersion, whereas all genes are used for DGE analysis.
> 
> ```r
> ### Copy normalized DGEList object (new object counts_object_1) and replace group variable with only one treatment group
> counts_object_1=counts_object
> counts_object_1$samples$group <- 1
> 
> ### Estimate common dispersion from a set of housekeeping genes (linked below). 
> y0 <- estimateDisp(counts_object_1[housekeeping,], trend="none", tagwise=FALSE)
> counts_object$common.dispersion <- y0$common.dispersion
> 
> ### Make design matrix 
> design <- model.matrix(~Groups)
> fit <- glmFit(counts_object, design)
> lrt <- glmLRT(fit)
> ```

Yes, you are right. It was used to estimate dispersions, in the special case of single-sample dge analysis referenced in the ticket. 

However for the purposes of this analysis, performing DGE on the subset of HRT genes (where subsetting is performed after filtering lowly expressed genes and estimating size factors but prior to dispersion estimation and DGE for each analysis) would allow for the identification of negative control genes that are identified statistically and are biologically known to be required for basic cellular functions. Sorry this was not clearly delineated in the ticket.

logstar commented:
> > @aadamk Thank you for the detailed review.
> > I understand the proposed modifications to run DGE analysis only on housekeeping genes, but I would like to confirm that this is requested in [PediatricOpenTargets/ticket-tracker#17 (comment)](https://github.com/PediatricOpenTargets/ticket-tracker/issues/17#issue-896875722). In the following Krutika's code, the housekeeping genes are used to estimate common dispersion, whereas all genes are used for DGE analysis.
> > ```r
> > ### Copy normalized DGEList object (new object counts_object_1) and replace group variable with only one treatment group
> > counts_object_1=counts_object
> > counts_object_1$samples$group <- 1
> > 
> > ### Estimate common dispersion from a set of housekeeping genes (linked below). 
> > y0 <- estimateDisp(counts_object_1[housekeeping,], trend="none", tagwise=FALSE)
> > counts_object$common.dispersion <- y0$common.dispersion
> > 
> > ### Make design matrix 
> > design <- model.matrix(~Groups)
> > fit <- glmFit(counts_object, design)
> > lrt <- glmLRT(fit)
> > ```
> 
> Yes, you are right. It was used to estimate dispersions, in the special case of single-sample dge analysis referenced in the ticket.
> 
> However for the purposes of this analysis, performing DGE on the subset of HRT genes (where subsetting is performed after filtering lowly expressed genes and estimating size factors but prior to dispersion estimation and DGE for each analysis) would allow for the identification of negative control genes that are identified statistically and are biologically known to be required for basic cellular functions. Sorry this was not clearly delineated in the ticket.

Got it. Thank you for the clarification. I will update the analysis accordingly to only use the housekeeping genes for DGE analysis.

logstar commented:
Following are the updates on this PR:

**1. Run DGE analysis on housekeeping genes only.**

Implemented the suggested DGE analysis procedure:

> 1. Filter out non-expressed genes on the full transcriptome
> 2. Correct for compositional biases on the full transcriptome (TMM, uqpgq2, or DESeq2's method).
> 3. Filter down to housekeeping genes from https://housekeeping.unicamp.br/?download (Full Housekeeping gene list, ~2000 genes)
> 4. Estimate dispersions
> 5. Run DGE and plot p-val histograms.

Code for subsetting housekeeping genes:

<https://github.com/logstar/OpenPedCan-analysis/blob/d847ab2659bbf7efcdef29d06a882dabda9d5242/analyses/rna-seq-protocol-dge/01-protocol-dge-seg.R#L100-L133> (The link does not show code snippet for some reason.)

```R
### build normalized DGEList
counts_object <- DGEList(counts = counts, group = group)
counts_object <- counts_object[
    filterByExpr(counts_object), , keep.lib.sizes=FALSE]
# read housekeeping genes
hkgenes <- read.csv(file.path('input', 'Housekeeping_GenesHuman.csv'),
                    sep = ";", stringsAsFactors = FALSE)
hkgenes <- unique(hkgenes$Gene.name)
hkgenes <- intersect(rownames(counts_object), hkgenes)


if (identical(norm_method, 'tmm')) {
    counts_object <- calcNormFactors(counts_object, method="TMM")
    norm_cnt_mat <- cpm(counts_object)


    counts_object <- counts_object[hkgenes, ]
    norm_cnt_mat <- norm_cnt_mat[hkgenes, ]
    round_norm_cnt_mat <- round(norm_cnt_mat)


    plot_nm_str <- 'TMM'
} else if (identical(norm_method, 'uqpgq2')) {
    counts_object <- DGEList(counts = uq.pgQ2(X = counts_object$counts),
                             group = group)
    counts_object <- counts_object[hkgenes, ]


    norm_cnt_mat <- counts_object$counts
    round_norm_cnt_mat <- round(norm_cnt_mat)
    counts_object <- calcNormFactors(counts_object, method="none")


    plot_nm_str <- 'UQ-pgQ2'
}
### Estimate dispersion
counts_object <- estimateDisp(
    counts_object, design=design, prior.df=NULL, trend.method='locfit',
    tagwise=TRUE)
```

<https://github.com/logstar/OpenPedCan-analysis/blob/d847ab2659bbf7efcdef29d06a882dabda9d5242/analyses/rna-seq-protocol-dge/02-deseq2-protocol-dge.R#L44-L65>

```R
round_cnt_mat <- round(counts_object$counts)
suppressMessages(
  dds <- DESeq2::DESeqDataSetFromMatrix(
    countData = round_cnt_mat,
    colData = data.frame(group),
    design = ~ group)
)
# Read housekeeping genes
hkgenes <- read.csv(file.path('input', 'Housekeeping_GenesHuman.csv'),
                    sep = ";", stringsAsFactors = FALSE)
hkgenes <- unique(hkgenes$Gene.name)
hkgenes <- intersect(rownames(counts_object), hkgenes)
# From DESeq2 documentation:
# DESeq2 performs a default analysis through the steps:
# - estimation of size factors: estimateSizeFactors
# - estimation of dispersion: estimateDispersions
# - Negative Binomial GLM fitting and Wald statistics: nbinomWaldTest
dds <- DESeq2::estimateSizeFactors(dds)
dds <- dds[hkgenes, ]
suppressMessages(
  dds <- DESeq2::estimateDispersions(dds)
)
```

**2. Removed the collapsing procedure of RNA-seq read count matrices.**

Use PediatricOpenTargets/OpenPedCan-analysis release v4 data `gene-counts-rsem-expected_count-collapsed.rds` for DGE analysis. `gene-counts-rsem-expected_count-collapsed.rds` is merged PBTA and KFNBL RNA-seq raw count matrix.

Previously, AlexsLemonade/OpenPBTA-analysis release v19-20210423 data, `pbta-gene-counts-rsem-expected_count.polya.rds` and `pbta-gene-counts-rsem-expected_count.stranded.rds`, are collapsed in this analysis module, and the collapsed read count matrices are used for DGE analysis.

Code for loading `gene-counts-rsem-expected_count-collapsed.rds`:

<https://github.com/logstar/OpenPedCan-analysis/blob/d847ab2659bbf7efcdef29d06a882dabda9d5242/analyses/rna-seq-protocol-dge/01-protocol-dge-seg.R#L66-L89>

```R
cnt_mat <- readRDS('../../data/gene-counts-rsem-expected_count-collapsed.rds')


# 6 samples have both poly-A and stranded RNA-seq libraries
# Index | Kids_First_Biospecimen_ID | sample_id | experimental_strategy | 
#   RNA_library | cohort
# -- | -- | -- | -- | -- | --
# 1 | BS_HE0WJRW6 | 7316-1455 | RNA-Seq | stranded | CBTN
# 2 | BS_HWGWYCY7 | 7316-1455 | RNA-Seq | poly-A | CBTN 
# 3 | BS_SHJA4MR0 | 7316-161 | RNA-Seq | stranded | CBTN 
# 4 | BS_X0XXN9BK | 7316-161 | RNA-Seq | poly-A | CBTN 
# 5 | BS_FN07P04C | 7316-255 | RNA-Seq | stranded | CBTN 
# 6 | BS_W4H1D4Y6 | 7316-255 | RNA-Seq | poly-A | CBTN
# 7 | BS_8QB4S4VA | 7316-536 | RNA-Seq | stranded | CBTN
# 8 | BS_QKT3TJVK | 7316-536 | RNA-Seq | poly-A | CBTN
# 9 | BS_7WM3MNZ0 | A16915 | RNA-Seq | poly-A | PNOC003
# 10 | BS_KABQQA0T | A16915 | RNA-Seq | stranded | PNOC003
# 11 | BS_68KX6A42 | A18777 | RNA-Seq | poly-A | PNOC003
# 12 | BS_D7XRFE0R | A18777 | RNA-Seq | stranded | PNOC003
counts <- cbind(
  cnt_mat[, c("BS_HE0WJRW6", "BS_SHJA4MR0", "BS_FN07P04C",
              "BS_8QB4S4VA", "BS_KABQQA0T", "BS_D7XRFE0R")],
  cnt_mat[, c("BS_HWGWYCY7", "BS_X0XXN9BK", "BS_W4H1D4Y6",
              "BS_QKT3TJVK", "BS_7WM3MNZ0", "BS_68KX6A42")]
)
```

https://github.com/logstar/OpenPedCan-analysis/blob/d847ab2659bbf7efcdef29d06a882dabda9d5242/analyses/rna-seq-protocol-dge/02-deseq2-protocol-dge.R#L11-L34

Same as `01-protocol-dge-seg.R#L66-L89`.

aadamk reviewed (approved):
Many thanks for this analysis @logstar. 
This looks complete and it will be a matter of selecting the optimal normalization + stat test combination (then the associated negative control gene set).

logstar commented:
@aadamk Thank you very much for reviewing the DGE analysis part of this analysis module.

If you need any additional information for comparing different normalization + stat test procedures, feel free to let me know.

kgaonkar6 commented:
Hi @logstar could you get the most updated origin/dev branch and merge it to your `rna-seq-protocol-dge-third` branch? 
I see that the github actions is failing and that the actual github action [workflow](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/.github/workflows)  in dev branch is not updated in your current branch.

logstar commented:
> Hi @logstar could you get the most updated origin/dev branch and merge it to your `rna-seq-protocol-dge-third` branch?
> I see that the github actions is failing and that the actual github action [workflow](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/.github/workflows) in dev branch is not updated in your current branch.

@kgaonkar6 Thank you for asking. I have not merged the most updated dev branch to this PR yet. I will merge shortly.

kgaonkar6 commented:
Thanks for the merge. But.. still seems to fail. I can make separate PR to update the Docker file and check if that passes or else I might need to update the Docker credentials attached to the repo.

logstar commented:
> Thanks for the merge. But.. still seems to fail. I can make separate PR to update the Docker file and check if that passes or else I might need to update the Docker credentials attached to the repo.

Thanks for checking. Is it because this PR is submitted from my own fork at <https://github.com/logstar/OpenPedCan-analysis/tree/rna-seq-protocol-dge-third>? The error message says "Error: Username and password required", so I guess my fork does not have `${{ secrets.DOCKER_HUB_USERNAME }}` or `${{ secrets.DOCKER_HUB_ACCESS_TOKEN }}`.

kgaonkar6 commented:
You are right!I guess because we added the docker SECRETS after you forked the repo, it was not carried over to your fork. 

I've added the DESeq2 to dockerfile and seems to pass the checks https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/19. I still have to think about this is in detail but I think moving forward for any Docker file update we would have to make a branch within this main repo add a PR to check if it will be built ok before pushing to `dev`

kgaonkar6 commented:
You can definitely just create build docker locally and push to dockerhub without using the github actions. By testing on Github actions it just saves time of an actual person who would otherwise test it locally.
I checked the Actions page and it says the github SECRETS which currently have my docker login credentials cannot be accessed if the PR is from a fork so any Dockerfile update will need to be a branch on this  main repo itself.

For this PR I think you can remove the Dockerfile update so there are no checks and we will be able to merge.

logstar commented:
> You can definitely just create build docker locally and push to dockerhub without using the github actions. By testing on Github actions it just saves time of an actual person who would otherwise test it locally.
> I checked the Actions page and it says the github SECRETS which currently have my docker login credentials cannot be accessed if the PR is from a fork so any Dockerfile update will need to be a branch on this main repo itself.
> 
> For this PR I think you can remove the Dockerfile update so there are no checks and we will be able to merge.

I see. Thank you for the suggestion. I have updated the Dockerfile to be the same as the dev branch.

kgaonkar6 commented on analyses/rna-seq-protocol-dge/01-protocol-dge-seg.R:
```diff
@@ -0,0 +1,323 @@
+lrt <- glmLRT(glm_fit)
+
+# Save LRT p-value table and histogram
+lrt_top_tags <- topTags(lrt, adjust.method = "BH", n = Inf)
+lrt_out_df <- lrt_top_tags$table
+stopifnot(identical(colnames(lrt_out_df),
+                    c("logFC", "logCPM", "LR", "PValue", "FDR")))
+colnames(lrt_out_df) <- c("stranded_over_polya_logFC",
+                          "average_logCPM", "LR", "PValue", "BH FDR")
+write.csv(lrt_out_df, file=file.path(table_outdir,
+                                     'stranded_vs_polya_dge_lrt_res.csv'))
+
+# plot and save p-value histogram
+p <- ggplot(lrt_top_tags$table, aes(x=PValue)) +
+    geom_histogram(binwidth = 0.05, center = 0.025) +
+    theme_classic() +
+    scale_x_continuous(expand = expand_scale(mult = c(0, 0.02))) +
+    scale_y_continuous(expand = expand_scale(mult = c(0, 0.05))) +
+    xlab(paste0('stranded vs poly-A RNA-seq DGE ', plot_nm_str,
+                ' LRT p-value')) +
+    ylab('Gene count') +
+    ggtitle(paste0('Histogram of stranded vs poly-A RNA-seq\n',
+                   'differential gene expression ', plot_nm_str,
+                   ' normalized\n',
+                   'LRT p-values\n',
+                   sum(lrt_top_tags$table$PValue < 0.05),
+                   ' genes have p-value < 0.05\n',
+                   sum(lrt_top_tags$table$PValue >= 0.05),
+                   ' genes have p-value >= 0.05\n',
+                   sum(lrt_top_tags$table$FDR < 0.05),
+                   ' genes have BH FDR < 0.05\n',
+                   sum(lrt_top_tags$table$FDR >= 0.05),
+                   ' genes have BH FDR >= 0.05')) +
+    theme(text = element_text(size=15))
+
+ggsave(file.path(plot_outdir, 'stranded_vs_polya_dge_lrt_pvals_histogram.png'),
+       dpi = 300, plot = p, width = 8, height = 7)
```
It seems to me that similar code is used for each test exactTest, glmLRT and glmQLFTest, could we add a function instead?

logstar replied:
Thank you for the suggestion. 

I have considered adding a function for the similar tabulating and plotting code, but I think it is more maintainable to have separate tabulating and plotting code for each test. When making changes to existing tests or adding new tests in the future, the intermediate results need to be checked carefully, and test-specific code may be necessary. For example, the column names of DESeq2 result table are different from edgeR, and the following specific code is added to handle the NAs in p-values. These specific differences may be overlooked when using a function that takes in the test result and various plotting labels. In addition, I think the tabulating and plotting procedures are simple enough for reuse in other scenarios.

```R
# plot and save p-value histogram
# some DESeq2 p-values are NAs due to all-0 normalized read counts
deseq2_plot_df <- deseq2_res_df[, c('pvalue', 'padj')]
deseq2_plot_df[is.na(deseq2_plot_df)] <- 1
```

kgaonkar6 replied:
Oh I suggested it just for the edgeR tests since they seem to be the exact same code.

kgaonkar6 replied:
It's totally not necessary (just thought of adding it as a suggestion) OK to merge without it.

logstar replied:
Sorry that I misunderstood your suggestion. I thought all tests were concerned.

Thank you for your understanding. I think the code might be good as is, as it is simple enough to read without any function. In the future updates, I will refactor the repeated tabulating and plotting procedures, if they become more complex.

kgaonkar6 reviewed (commented):
Great documentation! 

The scripts look good in general, my only comment was related to adding a function for the various edgeR test and plot code chunks.
